src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,270 @@
+    bool reverse_parameterization = false;
+
+    // Check that the pre conditions are met
+    assert(num_segments > 2);
+    assert(max_allowable_acceleration > 0);
+    assert(max_allowable_speed > 0);
+    assert(initial_speed >= 0);
+    assert(final_speed >= 0);
+
+    // Check for the parameterization direction
+    // If the path is traversed in reverse, then flip all components to forwards (to be
+    // reversed again in the end)
+    if (t_end < t_start)
+    {
+        reverse_parameterization = true;
+
+        // Reverse the direction (Polynomial library can only handle forwards direction)
+        float temp = t_start;
+        t_start    = t_end;
+        t_end      = temp;
+
+        // Reverse the final and initial speeds
+        temp          = initial_speed;
+        initial_speed = final_speed;
+        final_speed   = temp;
+    }
+
+    // Create the parmeterization to contain the desired number of segments
+    CREATE_STATIC_ARC_LENGTH_PARAMETRIZATION(arc_length_param,
+                                             TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS);
+
+    // Get all of the points for the arc length parameterization (Not constant arc length
+    // segments)
+    shared_polynomial_getArcLengthParametrizationOrder3(path, t_start, t_end,
+                                                        arc_length_param);
+
+    const float arc_segment_length =
+        app_trajectory_planner_get_total_arcLength(arc_length_param) / num_segments;
+
+    // Now use numerical interpolation to get constant arc length segments for the
+    // parameterization
+    app_trajectory_planner_get_max_allowable_speed_profile(
+        max_allowable_speed_profile, traj_elements, path, num_segments, arc_length_param,
+        arc_segment_length, max_allowable_acceleration);
+
+    // First point on the velocity profile is always the current speed
+    velocity_profile[0] = initial_speed;
+
+    // Loop through the path forwards to ensure that the maximum velocity limit defined by
+    // the curvature is not breached by constant max acceleration of the robot (if it is,
+    // pull down the acceleration)
+    for (unsigned int i = 1; i < num_segments; i++)
+    {
+        // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
+        const float temp_vel = sqrt(pow(velocity_profile[i - 1], 2) +
+                                    2 * arc_segment_length * max_allowable_acceleration);
+
+        // If the new speed is greater than the max allowable, reduce it to the max
+        if (temp_vel > max_allowable_speed_profile[i])
+        {
+            velocity_profile[i] = max_allowable_speed_profile[i];
+        }
+        // If the new speed is greater than the maximum limit of the robot, reduce it
+        // to the limit
+        else if (temp_vel >= max_allowable_speed)
+        {
+            velocity_profile[i] = max_allowable_speed;
+        }
+        else
+        {
+            velocity_profile[i] = temp_vel;
+        }
+    }
```

Comment:
Could we break this out into a function?

Reply:
Broke out into app_trajectory_planner_generate_forwards_velocity_profile()